### ISSUE TITLE
"no-unnecessary-type-assertion" false positive.

### ISSUE THREAD

**Issue by bowenni** [palantir/tslint#3730]
### Bug Report

- __TSLint version__: 5.9.1
- __TypeScript version__: 2.6.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
let foo: {};
const bar = () => {
  foo = {};
};
bar();
const baz = foo!;
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "no-unnecessary-type-assertion": true
    }
}
```

#### Actual behavior
`ERROR: myFile.ts[6, 13]: This assertion is unnecessary since it does not change the type of the expression.`

#### Expected behavior
No errors. If I remove `!` TypeScript compiler fails with `myFile.ts(6,13): error TS2454: Variable 'foo' is used before being assigned.`

**Comment by ajafff** [comment]
Duplicate of #3521
I have a working solution in `no-useless-assertion` of `@fimbul/wotan` https://github.com/fimbullinter/wotan/blob/master/packages/wotan/README.md#available-rules

**Comment by bowenni** [comment]
Thanks!
